### PR TITLE
Framework

### PR DESCRIPTION
**Pull request by ctn**
1. @nhanitvn : refactoring base into Model and Modeler, and associated subclasses (e.g., PredictiveModel, RuleBasedModel, BooleanModel, etc.)
2. @ctn : (re)introduced the requirement of design-and-communicate-via-code workflow, via the user/ directory. This must be used by Product to communicate client needs to framework implementors, and ahead of any implementation.
3. @ctn @TheVinhLuong102 : miscellaneous config changes to make poetry work properly with scikitlearn etc. on an M1 Mac.

### CODE REVIEW

**Comment by ctn** [comment]
All checks have passed :)

File: user/model/__init__.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Don’t forget to import __init__
```
Comment (nhanitvn) [inline user/model/__init__.py:1]: This is only needed where we could not use Poetry such as CTN's Apple M1 MacBook. TODO: investigate, fix Poetry issues for that case and remove this __init__.py file.

**Review by nhanitvn — APPROVED** [review]
lgtm

**Comment by ctn** [comment]
Move fast and yet not breaking things 😉

**Comment by ctn** [comment]
OK then imma merge.

File: user/model/__init__.py (hunk 1 above)
Comment (ctn) [reply user/model/__init__.py:1]: Yep.

Comment (ctn) [reply user/model/__init__.py:1]: Another workaround is to not develop on M1!

**Pull request merged**